plugin/restore/mysql/mysql.go

```diff
@@ -680,6 +680,221 @@ func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]Binlo
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	// s.Scan() could panic, so we should recover and handle this.
+	defer func() {
+		if r := recover(); r != nil {
+			var ok bool
+			err, ok = r.(error)
+			if !ok {
+				err = fmt.Errorf("%v", r)
+			}
+			log.Error("bufio.Scanner panic", zap.Error(err))
+		}
+	}()
+	var found bool
+	for s.Scan() {
+		line := s.Text()
+		eventTs, found, err = parseBinlogEventTsInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog eventTs from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+		break
+	}
+
+	return eventTs, nil
+}
+
+// Use command like mysqlbinlog --start-datetime=targetTs+1 binlog.000001 to parse the first binlog event ts after targetTs.
+// The --start-datetime is inclusive, so we use targetTs+1 as the argument, which gives the position after all events before targetTs (inclusive) are replayed.
+// TODO(dragonly): Optimize to kill mysqlbinlog process after we parsed the first binlog event ts after targetTs.
+// The current implementation will parse the entire binlog file. This optimization may significantly reduce CPU/memory/disk IO usage.
+// TODO(dragonly): Add integration test.
+func (r *Restore) getFirstBinlogEventStartPosAfterTs(ctx context.Context, binlogFile BinlogFile, targetTs int64) (startPos int64, err error) {
+	args := []string{
+		// Local binlog file path.
+		path.Join(r.binlogDir, binlogFile.Name),
+		// Tell mysqlbinlog to suppress the BINLOG statements for row events, which reduces the unneeded output.
+		"--base64-output=DECODE-ROWS",
+		// Instruct mysqlbinlog to start output only after encountering the first binlog event with timestamp equal or after targetTs+1.
+		"--start-datetime", formatDateTime(targetTs + 1),
+	}
+	cmd := exec.CommandContext(ctx, r.mysqlutil.GetPath(mysqlutil.MySQLBinlog), args...)
+	cmd.Stderr = os.Stderr
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return 0, err
+	}
+	s := bufio.NewScanner(pr)
+	if err := cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	// s.Scan() could panic, so we should recover and handle this.
+	defer func() {
+		if r := recover(); r != nil {
+			var ok bool
+			err, ok = r.(error)
+			if !ok {
+				err = fmt.Errorf("%v", r)
+			}
+			log.Error("bufio.Scanner panic", zap.Error(err))
+		}
+	}()
+
+	// When invoking mysqlbinlog with --start-datetime, the first valid event will always be the starting event, and the timestamp is when the binlog file is created.
+	// We must skip this event, and parse the next one.
+	foundEvent := 0
+	var found bool
+	for s.Scan() {
+		line := s.Text()
+		startPos, found, err = parseBinlogEventStartPosInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog event start position from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+
+		foundEvent++
+		if foundEvent == 2 {
```

Comment:
Why 2?

Reply:
There's a comment explaining this and I moved it near this line.
> // When invoking mysqlbinlog with --start-datetime, the first valid event will always be the starting event, and the timestamp is when the binlog file is created.
> // We must skip this event, and parse the next one.

Reply:
Changed to filter the `FORMAT_DESCRIPTION_EVENT` event at position 4.